test/utils.dart

```diff
@@ -25,17 +25,89 @@ const formattedOutput = 'void main() => print("hello");';
 final _indentPattern = RegExp(r'\(indent (\d+)\)');
 final _fixPattern = RegExp(r'\(fix ([a-x-]+)\)');
 
-/// Runs the command line formatter, passing it [args].
-Future<TestProcess> runFormatter([List<String> args]) {
+/// If tool/command_shell.dart has been compiled to a snapshot, this is the path
+/// to it.
+String _commandExecutablePath;
+
+/// If bin/format.dart has been compiled to a snapshot, this is the path to it.
+String _formatterExecutablePath;
+
+/// Compiles format.dart to a native executable for tests to use.
+void compileFormatterExecutable() {
+  setUpAll(() async {
+    _formatterExecutablePath = await _compileExecutable('bin/format.dart');
+  });
+
+  tearDownAll(() async {
+    await _deleteExecutable(_formatterExecutablePath);
+    _formatterExecutablePath = null;
+  });
+}
+
+/// Compiles command_shell.dart to a native executable for tests to use.
+void compileCommandExecutable() {
+  setUpAll(() async {
+    _commandExecutablePath =
+        await _compileExecutable('tool/command_shell.dart');
+  });
+
+  tearDownAll(() async {
+    await _deleteExecutable(_commandExecutablePath);
+    _commandExecutablePath = null;
+  });
+}
+
+/// Compile the Dart [script] to a native executable.
+///
+/// We do this instead of spawning the script from source each time because it's
+/// much faster when the same script needs to be run several times.
+Future<String> _compileExecutable(String script) async {
+  var scriptName = p.basename(script);
+  var tempDir =
+      await Directory.systemTemp.createTemp(p.withoutExtension(scriptName));
+  var executable = p.join(tempDir.path, '$scriptName.aot');
+
+  var dart2Native =
+      p.join(p.dirname(Platform.resolvedExecutable), 'dart2native');
+
   // Locate the "test" directory. Use mirrors so that this works with the test
   // package, which loads this suite into an isolate.
```

Comment:
[out of scope] I don't think you should need mirrors for this. `Isolate.resolvePackageUri` would let you find `lib/` and I think it's fair to assume relative path of `test/` from `lib/`, especially since we already assume a relative path of `bin/` from `test/`...

Reply:
I want to get this landed so I can roll into the SDK soon, so added a TODO to look into this later.